### PR TITLE
Tar support

### PR DESCRIPTION
**Pull request by nxpweilincao**

### CODE REVIEW

File: libuuu/buffer.cpp [hunk 1]
```diff
@@ -114,6 +115,8 @@ class FSBasic
 			*outfilename = filename.substr(pos + strlen(m_ext) + 1);
 		else
 			outfilename->clear();
+
+		cout<<"filename: "<<filename<<" outfilename: "<<*outfilename<<" outbackfile: "<<*outbackfile<<endl;
 		return 0;
```
Comment (nxpfrankli) [inline libuuu/buffer.cpp]: Remove debug code here

File: libuuu/buffer.cpp [hunk 2]
```diff
@@ -262,6 +275,8 @@ static class FS_DATA
 	{
 		m_pFs.push_back(&g_fsflat);
 		m_pFs.push_back(&g_fszip);
+		//
+		m_pFs.push_back(&g_fstar);
```
Comment (nxpfrankli) [inline libuuu/buffer.cpp]: remove empty comments line

File: libuuu/buffer.cpp [hunk 3]
```diff
@@ -702,12 +793,14 @@ int bz_async_load(string filename, shared_ptr<FileBuffer> p)
 
 int FSBz2::load(string backfile, string filename, shared_ptr<FileBuffer>p, bool async)
 {
+	
 	if (filename != "*")
```
Comment (nxpfrankli) [inline libuuu/buffer.cpp]: remove empty line

File: libuuu/buffer.cpp [hunk 4]
```diff
@@ -407,6 +422,82 @@ int FSZip::load(string backfile, string filename, shared_ptr<FileBuffer> p, bool
...
+			return -1;
+
+		atomic_fetch_or(&p->m_dataflags, FILEBUFFER_FLAG_LOADED);
+	}
+	return 0;
+}
+
+
+
+
```
Comment (nxpfrankli) [inline libuuu/buffer.cpp]: remove extra empty line

File: libuuu/buffer.cpp [hunk 5]
```diff
@@ -407,6 +422,82 @@ int FSZip::load(string backfile, string filename, shared_ptr<FileBuffer> p, bool
...
+		return -1;
+
+	if (!tar.check_file_exist(filename))
+		return -1;
+
+	if (async)
+	{
+		p->m_aync_thread = thread(tar_async_load, backfile, filename, p);
+	}
+	else
```
Comment (nxpfrankli) [inline libuuu/buffer.cpp]: if tar is quick enough.  We don't use async thread.

File: libuuu/buffer.cpp (hunk 3 above)
Comment (nxpfrankli) [inline libuuu/buffer.cpp]: remove extra empty line

File: libuuu/buffer.cpp [hunk 6]
```diff
@@ -702,12 +793,14 @@ int bz_async_load(string filename, shared_ptr<FileBuffer> p)
...
 {
+	
 	if (filename != "*")
 	{
 		set_last_err_string("bz just support . decompress itself");
 		return -1;
 	}
 
+
 	if (!g_fs_data.exist(backfile))
```
Comment (nxpfrankli) [inline libuuu/buffer.cpp]: extra empty line

File: libuuu/tar.h [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+#include <string.h>
+
+using namespace std;
+
+#define TAR_BLOCK_SIZE 512
+
+struct Tar_header
+{
+	char name[100];
+	char mode[8];
```
Comment (nxpfrankli) [inline libuuu/tar.h]: please use uint8_t for char

File: libuuu/tar.h [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+	char ustar[6];
+	char version[2];
+	char uname[32];
+	char gname[32];
+	char major_num[8];
+	char minor_num[8];
+	char prefix[155];
+};
+
+
```
Comment (nxpfrankli) [inline libuuu/tar.h]: Better added pack(1) for such structure

File: libuuu/tar.h [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+{
+	string filename;
+
+public:
+	map<string, Tar_file_Info> m_filemap;
+	int Open(string filename)
+	{
+		this->filename=filename;
+		ifstream stream(filename, ifstream::binary);
+		if(!stream)
```
Comment (nxpfrankli) [inline libuuu/tar.h]: use get_file_buffer(filename) should be simpler than ifstream.  Previous zip implement is not good enough.

File: msvc/bzip2.vcxproj [hunk 10]
```diff
@@ -91,32 +91,32 @@
...
-    <PlatformToolset>v141</PlatformToolset>
+    <PlatformToolset>v142</PlatformToolset>
     <CharacterSet>MultiByte</CharacterSet>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|x64'" Label="Configuration">
     <ConfigurationType>StaticLibrary</ConfigurationType>
     <UseDebugLibraries>false</UseDebugLibraries>
-    <PlatformToolset>v141</PlatformToolset>
+    <PlatformToolset>v142</PlatformToolset>
     <WholeProgramOptimization>true</WholeProgramOptimization>
```
Comment (nxpfrankli) [inline msvc/bzip2.vcxproj]: Does other project use V142?

Comment (nxpweilincao) [reply msvc/bzip2.vcxproj]: It seems like this is something the visual studio changed by itself, I did not do anything about this file. This is quite weird.

File: libuuu/buffer.cpp [hunk 11]
```diff
@@ -114,6 +115,7 @@ class FSBasic
 			*outfilename = filename.substr(pos + strlen(m_ext) + 1);
 		else
 			outfilename->clear();
+
```
Comment (nxpfrankli) [inline libuuu/buffer.cpp]: extra empty line

File: libuuu/buffer.cpp [hunk 12]
```diff
@@ -407,6 +420,78 @@ int FSZip::load(string backfile, string filename, shared_ptr<FileBuffer> p, bool
...
+int FSTar::load(string backfile, string filename, shared_ptr<FileBuffer> p, bool async)
+{
+	Tar tar;
+	if (tar.Open(backfile))
+		return -1;
+
+	if (!tar.check_file_exist(filename))
+		return -1;
+
+	if (async)
```
Comment (nxpfrankli) [inline libuuu/buffer.cpp]: Sync load should be enough

File: libuuu/tar.cpp [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+	atomic_fetch_or(&p->m_dataflags, FILEBUFFER_FLAG_KNOWN_SIZE);
+
+	shared_ptr<FileBuffer> file;
+	file = get_file_buffer(tarfilename);
+	size_t offset= m_filemap[filename].offset;
+	size_t size=m_filemap[filename].size;
+
+
+	p->resize(0);
+	//memcpy(p->data(), file->data()+offset,size );
```
Comment (nxpfrankli) [inline libuuu/tar.cpp]: p->resize(0) not necessary

File: libuuu/tar.h [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+	uint64_t offset;
+	uint64_t size;
+
+};
+
+
+class Tar
+{
+	string tarfilename;
+
```
Comment (nxpfrankli) [inline libuuu/tar.h]: tarfilename -> m_tarfilename

File: libuuu/tar.h [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+	uint8_t ustar[6];
+	uint8_t version[2];
+	uint8_t uname[32];
+	uint8_t gname[32];
+	uint8_t major_num[8];
+	uint8_t minor_num[8];
+	uint8_t prefix[155];
+};
+
+
```
Comment (nxpfrankli) [inline libuuu/tar.h]: #pragma pack() to set back default value.

File: libuuu/tar.cpp [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+
+	p->resize(m_filemap[filename].size);
+	atomic_fetch_or(&p->m_dataflags, FILEBUFFER_FLAG_KNOWN_SIZE);
+
+	shared_ptr<FileBuffer> file;
+	file = get_file_buffer(m_tarfilename);
+	size_t offset= m_filemap[filename].offset;
+	size_t size=m_filemap[filename].size;
+
+	//memcpy(p->data(), file->data()+offset,size );
```
Comment (nxpfrankli) [inline libuuu/tar.cpp]: remove comments dead code

**Pull request merged**